test/test_charconv.cpp

```diff
@@ -20,93 +21,369 @@
+        CHECK_EQ(digits_dec<T>(999999999999999), 15u);
+        CHECK_EQ(digits_dec<T>(1000000000000000), 16u);
+        CHECK_EQ(digits_dec<T>(1000000000000001), 16u);
+        CHECK_EQ(digits_dec<T>(9999999999999999), 16u);
+        CHECK_EQ(digits_dec<T>(10000000000000000), 17u);
+        CHECK_EQ(digits_dec<T>(10000000000000001), 17u);
+        CHECK_EQ(digits_dec<T>(99999999999999999), 17u);
+        CHECK_EQ(digits_dec<T>(100000000000000000), 18u);
+        CHECK_EQ(digits_dec<T>(100000000000000001), 18u);
+        CHECK_EQ(digits_dec<T>(999999999999999999), 18u);
+    }
+}
+
+TEST_CASE_TEMPLATE("digits_bin", T, int8_t, uint8_t, int16_t, uint16_t, int32_t, uint32_t, int64_t, uint64_t)
+{
+    CHECK_EQ(digits_bin<T>(0x0), csubstr("0b0").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x1), csubstr("0b1").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x2), csubstr("0b10").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x3), csubstr("0b11").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x4), csubstr("0b100").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x5), csubstr("0b101").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x6), csubstr("0b110").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x7), csubstr("0b111").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x8), csubstr("0b1000").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x9), csubstr("0b1001").len - 2u);
+    CHECK_EQ(digits_bin<T>(0xa), csubstr("0b1010").len - 2u);
+    CHECK_EQ(digits_bin<T>(0xb), csubstr("0b1011").len - 2u);
+    CHECK_EQ(digits_bin<T>(0xc), csubstr("0b1100").len - 2u);
+    CHECK_EQ(digits_bin<T>(0xd), csubstr("0b1101").len - 2u);
+    CHECK_EQ(digits_bin<T>(0xe), csubstr("0b1110").len - 2u);
+    CHECK_EQ(digits_bin<T>(0xf), csubstr("0b1111").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x10), csubstr("0b10000").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x11), csubstr("0b10001").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x12), csubstr("0b10010").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x13), csubstr("0b10011").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x14), csubstr("0b10100").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x15), csubstr("0b10101").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x16), csubstr("0b10110").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x17), csubstr("0b10111").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x18), csubstr("0b11000").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x19), csubstr("0b11001").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x1a), csubstr("0b11010").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x1b), csubstr("0b11011").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x1c), csubstr("0b11100").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x1d), csubstr("0b11101").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x1e), csubstr("0b11110").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x1f), csubstr("0b11111").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x7e), csubstr("0b1111110").len - 2u);
+    CHECK_EQ(digits_bin<T>(0x7f), csubstr("0b1111111").len - 2u);
+    if(std::is_unsigned<T>::value)
+    {
+        //CHECK_EQ(digits_bin<T>(0xfe), 2u);
```

Comment:
Thanks for the thoroughness, it is really appreciated. To clarify, what (I (think) is finished as stated above is the R&D for picking the best combination of components in the algorithm for `write_dec()` in `bm_xtoa.cpp`. The test  is still a WIP, and I haven't touched it since starting with the investigation. Now I will implement the best method in `charconv.hpp` and fix the tests.